vllm/model_executor/models/utils.py

```diff
@@ -12,55 +12,167 @@
 
+WeightsMapping = Mapping[str, Optional[str]]
+"""If a key maps to a value of `None`, the corresponding weight is ignored."""
 
-class WeightsGroup(UserDict):
-    """
-    Wraps grouped weights dictionary for a more informative error message
-    when attempting to access a weight component that does not exist.
-    """
 
-    def __getitem__(self, key: str) -> Iterable[Tuple[str, torch.Tensor]]:
-        try:
-            return super().__getitem__(key)
-        except KeyError as exc:
-            msg = (f"There is no weights named with the prefix: {key}. "
-                   f"Available prefix: {set(self.keys())}")
-            raise KeyError(msg) from exc
+@dataclass
+class WeightsMapper:
+    """Maps the name of each weight if they match the following patterns."""
 
+    orig_to_new_substr: WeightsMapping = field(default_factory=dict)
+    orig_to_new_prefix: WeightsMapping = field(default_factory=dict)
+    orig_to_new_suffix: WeightsMapping = field(default_factory=dict)
 
-def filter_weights(weights: Iterable[Tuple[str, torch.Tensor]],
-                   prefix: str) -> Iterable[Tuple[str, torch.Tensor]]:
-    """
-    Helper function to load weights for inner vLLM models.
+    def apply(self, key: str) -> Optional[str]:
+        for substr, new_key in self.orig_to_new_substr.items():
+            if substr in key:
+                if new_key is None:
+                    return None
 
-    See also:
-        :ref:`init_vllm_registered_model`
-    """
-    for name, loaded_weight in weights:
-        name = name.split(".")
-        if prefix == name.pop(0):
-            name = ".".join(name)
-            yield name, loaded_weight
+                key = key.replace(substr, new_key, 1)
+
+        for prefix, new_key in self.orig_to_new_prefix.items():
+            if key.startswith(prefix):
+                if new_key is None:
+                    return None
+
+                key = key.replace(prefix, new_key, 1)
+
+        for suffix, new_key in self.orig_to_new_suffix.items():
+            if key.endswith(suffix):
+                if new_key is None:
+                    return None
 
+                key = new_key.join(key.rsplit(suffix, 1))
 
-def group_weights_with_prefix(
-    weights: Iterable[Tuple[str, torch.Tensor]], ) -> WeightsGroup:
+        return key
+
+
+class AutoWeightsLoader:
     """
-    Helper function to group weights with prefix
+    Helper class to load weights into a :class:`torch.nn.Module`. It is able
+    to automatically detect child modules and parameters while iterating over
+    the weights only once.
+
+    The weight loading logic for individual modules can be overridden
+    by defining a ``load_weights`` method.
+
+    Similarly, the weight loading logic for individual parameters can be
+    overridden by defining a ``weight_loader`` method.
     """
-    init_weights, repeated_weights = itertools.tee(weights, 2)
-    weights_prefix = {name.split(".")[0] for name, _ in init_weights}
```

Comment:
Ooops, I found here is the root cause of the memory overhead, because this step is not lazy!